### PR TITLE
Update Socket component to v0.5

### PR DESCRIPTION
**Pull request by clue**
Filing this as a WIP until the v0.5 release has actually been tagged and released.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "require": {
         "php": ">=5.4.0",
         "guzzlehttp/psr7": "^1.0",
-        "react/socket": "^0.4",
+        "react/socket": "dev-master as 0.5",
```
Comment (clue) [inline composer.json]: This should be `^0.5` once the release is out.

**Comment by clue** [comment]
Updated and rebased now that the Socket v0.5.0 release is out :shipit:

**Comment by WyriHaximus** [comment]
Looks good but I'm unsure about the `v0.4.5` milestone. Anyone updating to `v0.4.5` could break their code due to the changed socket API.

**Comment by jsor** [comment]
Oh yes, missed the milestone. Agree with @WyriHaximus.

**Comment by clue** [comment]
The HTTP component does not contain a breaking change and we do not expose any of the Socket API, so this is in fact entirely compatible.

Consumers of this package do in fact have to pass an instance of `SocketInterface` to the `Server`, so THEY have a dependency on the Socket component. If they stick to Socket v0.4, then they should not be able to update HTTP to v0.4.5. If they update their API to Socket v0.5, then they will receive all the shiny new features :+1:

**Comment by jsor** [comment]
Not sure i can agree. If someone gets the react/socket dependency only indirectly through react/http, it is a breaking change them.

**Comment by clue** [comment]
"If someone gets the react/socket dependency only indirectly through react/http," their code is already broken :-) If anybody uses `new React\Socket\Server(…)`, they DO in fact already depend on the Socket API. If they do so, then they will NOT be affected at all.

That being said, please review the above changes: For this most part, this only modifies our examples and test suite, there's only really a single change in the code (`getRemoteAddress()` now also contains the port). In other words: We only *consume* the Socket API, we do *not* expose its API.

**Comment by clue** [comment]
We COULD support both Socket v0.4 and v0.5 because the only BC issue we're affected by is the above `getRemoteAddress()` change which could easily be worked around so we support both formats.

However, for our test suite / examples we would have to pick one code path, which SHOULD be the most up-to-date version of the Socket v0.5 API (as per this PR), thus leaving the v0.4 code path effectively untested.

Besides, assuming we target `react/socket: ^0.5 || ^0.4` here, this could still break consumers that do not explicitly specify their Socket version.

Long story short: I'm not to decide this, but I'd rather ask (and help) consumers to fix *their* code than to block progress here and fix symptoms (proactively) here :+1:

**Comment by jsor** [comment]
Not sure i'm missing something, but if i now require `react/http:^0.4.4`, create a script with our [example code](https://github.com/reactphp/http/blob/master/examples/01-hello-world.php) and then do `composer update` after `0.4.5` is released, the script is broken. 🤔

**Comment by clue** [comment]
Correct, your assumed script would be broken, because you've missed to specify correct EventLoop and Socket dependencies :+1: 

In other words: You SHOULD already have specified this anyway, regardless of what Socket API we target here.

Also keep in mind that the fix here is rather trivial and we should emphasize this:
a) Fast fix only: Add `react/socket: ^0.4` dependency and do not upgrade to newest HTTP features
b) Looking foward: Make own code compatible with new socket API and add `react/socket: ^0.5`.

Note that we never run into any dependency problems (such as incompatible versions) here, as we've correctly specified all required versions. (not to be confused with dependency hell)

**Comment by clue** [comment]
TLDR: We've discussed this matter further via IRC and I agree it's safe to move this to the v0.5.0 milestone. :+1: 

While I still think it's reasonable for us to *not* consider this a breaking change, I agree that it's probably safer from a consumer perspective to do bump the major version regardless. This way consumers of this package will have to make an explicit, conscious switch to the newer Socket API.

I'd like to get this release out rather sooner than later in order to avoid working with the legacy API here. As such, I'm reorganizing some of the milestones to arrange for this :shipit:
